src/deluge/modulation/arpeggiator.cpp

```diff
@@ -772,32 +826,27 @@ bool ArpeggiatorForDrum::hasAnyInputNotesActive() {
 // Check arpeggiator is on before you call this.
 // May switch notes on and/or off.
 void ArpeggiatorBase::render(ArpeggiatorSettings* settings, int32_t numSamples, uint32_t gateThreshold,
-                             uint32_t phaseIncrement, uint32_t sequenceLength, uint32_t ratchAmount, uint32_t ratchProb,
+                             uint32_t phaseIncrement, uint32_t sequenceLength, uint32_t ratchAm, uint32_t ratchProb,
                              ArpReturnInstruction* instruction) {
 	if (settings->mode == ArpMode::OFF || !hasAnyInputNotesActive()) {
+		// Save some CPU by an early return if arp is off or no notes are active
 		return;
 	}
 
-	// Update live Sequence Length value with the most up to date value from automation
-	maxSequenceLength = (((int64_t)sequenceLength) * kMaxMenuValue + 2147483648) >> 32; // in the range 0-50
+	// Update live values from automation
+	{
+		// Update live Gate value with the most up to date value from automation
+		gate = gateThreshold >> 8;
 
-	// Update live ratchetProbability value with the most up to date value from automation
-	ratchetProbability = ratchProb >> 16; // just 16 bits is enough resolution for probability
+		// Update live Sequence Length value with the most up to date value from automation
+		maxSequenceLength = (((int64_t)sequenceLength) * kMaxMenuValue + 2147483648) >> 32; // in the range 0-50
 
-	// Update live ratchetAmount value with the most up to date value from automation
-	// Convert ratchAmount to either 0, 1, 2 or 3 (equivalent to a number of ratchets: OFF, 2, 4, 8)
-	uint16_t amount = ratchAmount >> 16;
-	if (amount > 45874) { // > 35 -> 50
-		ratchetAmount = 3;
-	}
-	else if (amount > 26214) { // > 20 -> 34
-		ratchetAmount = 2;
-	}
-	else if (amount > 6553) { // > 5 -> 19
-		ratchetAmount = 1;
-	}
-	else { // > 0 -> 4
-		ratchetAmount = 0;
+		// Update live ratchetProbability value with the most up to date value from automation
+		ratchetProbability = ratchProb >> 16; // just 16 bits is enough resolution for probability
+
+		// Update live ratchetAmount value with the most up to date value from automation
+		// Convert ratchAmount to either 0, 1, 2 or 3 (equivalent to a number of ratchets: OFF, 2, 4, 8)
+		ratchetAmount = ratchAm >> 16;
```

Comment:
Can you fix this bug seperately so it can be cherry picked? I don't think the rest of these additions should be moved across, we should aim to generalize the ability to modulate things pre note on with random and LFO1 instead.

Reply:
Hmmm I can't find the bug you mention. This seems fine. I am using the provided ratchAmount function parameter to convert it to discrete values of ratchetAmount
```

void ArpeggiatorBase::render(ArpeggiatorSettings* settings, int32_t numSamples, uint32_t gateThreshold,
                             uint32_t phaseIncrement, uint32_t sequenceLength, **uint32_t ratchAmount**, uint32_t ratchProb,
                             ArpReturnInstruction* instruction) {
	if (settings->mode == ArpMode::OFF || !hasAnyInputNotesActive()) {
		return;
	}

	// Update live Sequence Length value with the most up to date value from automation
	maxSequenceLength = (((int64_t)sequenceLength) * kMaxMenuValue + 2147483648) >> 32; // in the range 0-50

	// Update live ratchetProbability value with the most up to date value from automation
	ratchetProbability = ratchProb >> 16; // just 16 bits is enough resolution for probability

	// Update live ratchetAmount value with the most up to date value from automation
	// Convert ratchAmount to either 0, 1, 2 or 3 (equivalent to a number of ratchets: OFF, 2, 4, 8)
	**uint16_t amount = ratchAmount >> 16;
	if (amount > 45874) { // > 35 -> 50
		ratchetAmount = 3;
	}
	else if (amount > 26214) { // > 20 -> 34
		ratchetAmount = 2;
	}
	else if (amount > 6553) { // > 5 -> 19
		ratchetAmount = 1;
	}
	else { // > 0 -> 4
		ratchetAmount = 0;
	}**
```